### ISSUE TITLE
nixos-anywhere does not export ZFS pools before rebooting

### ISSUE THREAD

**Issue by nicoonoclaste** [numtide/nixos-anywhere#156]
**Describe the bug**

ZFS pools are (apparently) not exported prior to reboot, and may fail to import on first boot.

**To Reproduce**

1. Use a disko config with a ZFS pool imported during early boot, like `<disko/example/zfs.nix>`
2. Set `boot.zfs.forceImportRoot = false;` in the NixOS config.  This is “highly recommended” by NixOS' own docs
3. Run `nixos-anywhere $target`

**Observed behaviour**

System installs successfully, reboots, then initramfs fails to import the zpool.

**Expected behavior**

Successful import and boot.

**System information**

Current `nixos-anywhere` HEAD (e516d378d6a)
Target was a Lenovo x230, booted on the current NixOS minimal install image (nixos-minimal-23.05.1950.fcc147b1e93-x86_64-linux.iso)

**Additional context**

Things work properly when running `nixos-anywhere --no-reboot` and manually running `zpool export -a` prior to rebooting.

**Comment by Lassulus** [comment]
I guess we can try to run zfs export -a everytime before the final reboot?

**Comment by Lassulus** [comment]
https://github.com/numtide/nixos-anywhere/pull/157

**Comment by Mic92** [comment]
The disko pr was applied. Can we close this?

**Comment by ethnt** [comment]
I currently still have this issue, despite that PR being merged into disko. Even with `forceImportRoot` set to `true`, I cannot import the ZFS root pool on boot.

**Comment by Mic92** [comment]
Also you only ever need forceImportRoot set to false for network-based block devices like iscsi. Otherwise your local nvme never gets accidentally used twice. iscsi was also the primary use case of this check.

**Comment by Mic92** [comment]
> I currently still have this issue, despite that PR being merged into disko. Even with `forceImportRoot` set to `true`, I cannot import the ZFS root pool on boot.

That's a different issue than. Check that your hardware-configuration.nix has all required kernel modules to set up your block device driver, i.e. the nvme driver for nvme.

**Comment by phaer** [comment]
@nbraud Curious if https://github.com/nix-community/nixos-anywhere/pull/239 changes anything here?

**Comment by munnik** [comment]
I also ran into this problem and don't know how to solve it.

My disko config:
```nix
{ 
  inputs, 
  lib, 
  ... 
}:
{
  imports = [ inputs.disko.nixosModules.disko ];
  
  # hostid needs to be fetched before installing
  # ssh into machine and run hostid
  networking.hostId = "8425e349";

  disko.devices = {
    disk = {
      vda = {
        type = "disk";
        device = "/dev/vda";
        content = {
          type = "gpt";
          partitions = {
            boot = {
              size = "1M";
              type = "EF02"; # for grub MBR
            };
            ESP = {
              size = "64M";
              type = "EF00";
              content = {
                type = "filesystem";
                format = "vfat";
                mountpoint = "/boot";
                mountOptions = [ "umask=0077" ];
              };
            };
            zfs = {
              size = "100%";
              content = {
                type = "zfs";
                pool = "zroot";
              };
            };
          };
        };
      };
    };
    zpool = {
      zroot = {
        type = "zpool";
        mode = "";
        options = {
          ashift = "12";
          autotrim = "on";
          autoexpand = "on";
          cachefile = "none";
        };        
        rootFsOptions = {
          compression = "zstd";
          mountpoint = "none";
          atime = "off";
          xattr = "sa";
          canmount = "off";
          devices = "off";
        };
        postCreateHook = "zfs list -t snapshot -H -o name | grep -E '^zroot@blank$' || zfs snapshot zroot@blank";
        datasets = {
          root = {
            type = "zfs_fs";
            mountpoint = "/";
          };
          nix = {
            type = "zfs_fs";
            mountpoint = "/nix";
          };
          var = {
            type = "zfs_fs";
            mountpoint = "/var";
          };
          home = {
            type = "zfs_fs";
            mountpoint = "/home";
          };
        };
      };
    };
  };
}
```

I made sure I'm running the latest version of disko and nixos-anywhere by running `nix flake update`. To install the machine I run `nix run github:nix-community/nixos-anywhere -- --flake .#vm-node --target-host root@192.168.122.43 --gene
rate-hardware-config nixos-facter ./facter.json`. Output of the (last part) of the installation:

```
### Installing NixOS ###
Pseudo-terminal will not be allocated because stdin is not a terminal.
Warning: Permanently added '192.168.122.53' (ED25519) to the list of known hosts.
installing the boot loader...
setting up /etc...
generating machine-specific age key...
Public key: age133s623lr9utv3f6nxteclz97yf3e0f4hmjncag3l2uxsmq73wevqv63rsc
setting up secrets...
Cannot read ssh key '/etc/ssh/ssh_host_rsa_key': open /etc/ssh/ssh_host_rsa_key: no such file or directory
Cannot read ssh key '/etc/ssh/ssh_host_ed25519_key': open /etc/ssh/ssh_host_ed25519_key: no such file or directory
/nix/store/1257mm0yrrvzvyzikwzbw1c2h4ixyf0c-sops-install-secrets-0.0.1/bin/sops-install-secrets: failed to decrypt '/nix/store/jqvnga6b5lcz92b22047s5p3hbvz8qm6-secrets.yaml': Error getting data key: 0 successful groups required, got 0
Activation script snippet 'setupSecrets' failed (1)
updating GRUB 2 menu...
installing the GRUB 2 boot loader on /dev/vda...
Installing for i386-pc platform.
Installation finished. No error reported.
installation finished!
umount: /mnt/var (zroot/var) unmounted
umount: /mnt/nix (zroot/nix) unmounted
umount: /mnt/home (zroot/home) unmounted
umount: /mnt/boot unmounted
umount: /mnt (zroot/root) unmounted
### Waiting for the machine to become unreachable due to reboot ###
Warning: Permanently added '192.168.122.53' (ED25519) to the list of known hosts.
### Done! ###
```

Screenshots of the VM while booting, importing the root zpool takes a while. I *think* it fails and then the next errors appear.

![Image](https://github.com/user-attachments/assets/4de9aa3e-638d-4ced-a535-37335222d2f4)
![Image](https://github.com/user-attachments/assets/47d55c5a-938c-483f-9b44-7450ba73aafe)

**Comment by munnik** [comment]
Running `nix run github:nix-community/nixos-anywhere -- --flake .#vm-node --target-host root@192.168.122.53 --gene
rate-hardware-config nixos-facter ./facter.json --phases kexec,disko,install` and then ssh to the machine and run:
```
[root@nixos-installer:~]# zpool list -v
NAME             SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
zroot            124G  1.10G   123G        -         -     0%     0%  1.00x    ONLINE  /mnt
  disk-vda-zfs   125G  1.10G   123G        -         -     0%  0.88%      -    ONLINE

[root@nixos-installer:~]# zpool status
  pool: zroot
 state: ONLINE
status: Some supported and requested features are not enabled on the pool.
        The pool can still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(7) for details.
config:

        NAME            STATE     READ WRITE CKSUM
        zroot           ONLINE       0     0     0
          disk-vda-zfs  ONLINE       0     0     0

errors: No known data errors

[root@nixos-installer:~]# zfs list
NAME         USED  AVAIL  REFER  MOUNTPOINT
zroot       1.10G   119G    96K  none
zroot/home    96K   119G    96K  /mnt/home
zroot/nix   1.10G   119G  1.10G  /mnt/nix
zroot/root   280K   119G   280K  /mnt
zroot/var    164K   119G   164K  /mnt/var

[root@nixos-installer:~]# hostid
8425e349

[root@nixos-installer:~]# umount /mnt/boot

[root@nixos-installer:~]# zpool export zroot

[root@nixos-installer:~]# zpool list -v
no pools available

[root@nixos-installer:~]# reboot
```
gives the same result...

**Comment by munnik** [comment]
https://discourse.nixos.org/t/cannot-import-zfs-pool-at-boot/4805

The provided solution also worked for me!

```nix
boot.zfs.devNodes = "/dev/disk/by-path";
```
